### DISCUSSION TITLE
Skip field in array of objects

### DISCUSSION THREAD

**Discussion by bombillazo** [RebeccaStevens/deepmerge-ts#163] [Q&A]
Hey! So far skipping fields in objects at the top level has been very easy, but I'm trying to skip fields of objects inside an array and cant figure out how to. How would one go about doing this using deepmerge?

**Comment by RebeccaStevens** [comment]
Could you give a small example of 2 things you want to merge and the expected output?

**Reply by bombillazo** [reply]
```ts
const remoteData = {
  data: [
    {
      foo: 'foo1',
      bar: 'bar1',
      baz: 'baz1'
    },
    {
      foo: 'foo2',
      bar: 'bar2',
      baz: 'baz2'
    },
  ]
}

const template = {
  data: [
    {
      foo: '',
      bar: '',
    },
    {
      foo: '',
      bar: '',
    },
  ]
}
```

I want to skip `baz` in the merged object.

**Reply by RebeccaStevens** [reply]
What should the output look like?
Are the arrays being concatenated (the default behavior) or are they being merge in another way (such as by index)?

**Reply by bombillazo** [reply]
Let me edit the original and show how the output should look like:

```ts
const output = {
  data: [
    {
      foo: 'foo1',
      bar: 'bar1',
    },
    {
      foo: 'foo2',
      bar: 'bar2',
    },
  ]
}
```

And thank you for the response/attention.

**Reply by RebeccaStevens** [reply]
The function `recordTemplateMerge` below will merge records the way you want.
The function `arrayTemplateMerge` will deconstruct the arrays you have and pass them to that merge function for you.

`const result = arrayTemplateMerge(template, remoteData);` should give the result you want.

```ts
import { deepmerge, deepmergeCustom } from "deepmerge-ts";
import { transpose  } from "./utils";

const recordTemplateMerge = deepmergeCustom({
  mergeRecords: (values) => {
    const merged = deepmerge(...values) as Record<PropertyKey, unknown>;
    const template = values[0];

    const keysToKeep = new Set<PropertyKey>([
      ...Object.keys(template),
      ...Object.getOwnPropertySymbols(template)
    ]);

    const allKeys: PropertyKey[] = [
      ...Object.keys(merged),
      ...Object.getOwnPropertySymbols(merged)
    ];

    for (const key of allKeys) {
      if (!keysToKeep.has(key)) {
        delete merged[key];
      }
    }

    return merged;
  }
});

const arrayTemplateMerge = deepmergeCustom({
  mergeArrays: (values) => {
    return transpose(values).map((v) => recordTemplateMerge(...v));
  }
});
```

Here's the transpose function:

```ts
export const transpose = (value: readonly (readonly unknown[])[]) =>
  value.length === 0
    ? []
    : value[0]!.map((_, colIndex) => value.map((row) => row[colIndex]));
```

Note:

The two functions `recordTemplateMerge` and `arrayTemplateMerge` could easily be made one but that would then apply the template merging to all records, not just the ones in the array.

**Reply by bombillazo** [reply]
Thanks for the help! I'll experiment with this, I got so far using the `mergeOthers:` inner function and ideally till be done in a single customMerge.
